girderformindlogger/api/v1/applet.py

```diff
@@ -170,24 +173,31 @@ def assignGroup(self, folder, group, role, subject):
     )
     def createApplet(self, protocolUrl=None, name=None, informant=None):
         thisUser = self.getCurrentUser()
-        thread = threading.Thread(
-            target=AppletModel().createAppletFromUrl,
-            kwargs={
-                'name': name,
-                'protocolUrl': protocolUrl,
-                'user': thisUser,
-                'constraints': {
-                    'informantRelationship': informant
-                } if informant is not None else None
-            }
-        )
-        thread.start()
-        return({
-            "message": "The applet is being created. Please check back in "
-                       "several mintutes to see it. If you have an email "
-                       "address associated with your account, you will receive "
-                       "an email when your applet is ready."
+        existingApplet = AppletModel().findOne({
+            'name': name
         })
+        if existingApplet is None:
+            thread = threading.Thread(
+                target=AppletModel().createAppletFromUrl,
+                kwargs={
+                    'name': name,
+                    'protocolUrl': protocolUrl,
+                    'user': thisUser,
+                    'constraints': {
+                        'informantRelationship': informant
+                    } if informant is not None else None
+                }
+            )
+            thread.start()
+            return({
+                "message": "The applet is being created. Please check back in "
+                        "several mintutes to see it. If you have an email "
+                        "address associated with your account, you will receive "
+                        "an email when your applet is ready."
+            })
+        else:
+            raise ValidationException('An applet with that name already exists.', 'name')
```

Comment:
This code block appears to be the meat of attempting to resolve #609. 

However, most cases will not be caught by this check.

If the optional param `name` is not provided to https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/api/v1/applet.py#L174

`name` will be inferred from the protocol 

https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/models/applet.py#L163-L166

and if an applet with that system name already exists, the new one will be automatically renamed with a number in parentheses following the name.

https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/models/applet.py#L91

Finally, there are several semantically keyed names that are preferred to the system name, and this function does not check against any of them:

https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/constants.py#L104

For example, the "MindLogger Demo" shown in this screenshot has a system `name` of "reprolib:protocols/mindlogger-demo/mindlogger-demo_schema (2)" but `skos:prefLabel` of "MindLogger Demo"

![admin screenshot](https://user-images.githubusercontent.com/5974438/74564307-c3cf5b80-4f3c-11ea-9b54-caefe0a5e631.png)

If I try to create a new "MindLogger Demo", this code lets me. 

```REST
POST /applet?protocolUrl=https%3A%2F%2Fraw.githubusercontent.com%2FReproNim%2Freproschema%2Fmaster%2Fprotocols%2Fmindlogger-demo%2Fmindlogger-demo_schema&name=MindLogger%20Demo
```
`200`
```json
{
  "message": "The applet is being created. Please check back in several mintutes to see it. If you have an email address associated with your account, you will receive an email when your applet is ready."
}
```

If I try to create a new "reprolib:protocols/mindlogger-demo/mindlogger-demo_schema (2)", the code correctly rejects.

```REST
POST /applet?protocolUrl=https%3A%2F%2Fraw.githubusercontent.com%2FReproNim%2Freproschema%2Fmaster%2Fprotocols%2Fmindlogger-demo%2Fmindlogger-demo_schema&name=reprolib%3Aprotocols%2Fmindlogger-demo%2Fmindlogger-demo_schema%20(2)
```
`400`
```json
{
  "field": "name",
  "message": "An applet with that name already exists.",
  "type": "validation"
}
```

Reply:
Hi, @shnizzedy. I have a few questions.
- If I input protocol URL already used, creation of the applet should be blocked?
- If name field is not specified, Protocol().preferredName(protocol) will be the applet name.
   If it duplicates, what actions should be taken?

Reply:
The issue says

> To resolve this issue, 
> 1. [ ] design a strategy for
>     - [ ] checking for any existing applets with the same name as an applet being created or renamed
>     - [ ] if an applet with that name already exists, renaming the new applet without giving the creator of the new applet details of the existing same-name applet
> 2. [ ] implement that strategy

So @webtalenttop, to answer your questions:

> If I input protocol URL already used, creation of the applet should be blocked?

Multiple applets should be able to use the same protocol, so at the very least a user reusing a protocol should have the opportunity to choose a new name. I generally lean toward automation, (eg, "Existing Name (2)" or using something fun like [anonymous Google Docs animals](https://medium.com/@roeder/these-are-all-of-the-anonymous-google-docs-animals-i-have-seen-so-far-8e33e0405993) ("Existing Name: Alligator"))

> If name field is not specified, Protocol().preferredName(protocol) will be the applet name. If it duplicates, what actions should be taken?

The same strategy as whatever you decide for above, ie,  at the very least a user reusing a name should have the opportunity to choose a new name, or an automated renaming strategy.

See [Schema Standardization](https://github.com/ReproNim/reproschema#schema-standardization) for more about the motivations and implementation of protocols in this project.